### ISSUE TITLE
evil-scroll-line-up / -down, possible improvements

### ISSUE THREAD

**Issue by duianto** [emacs-evil/evil#778]
### Description
Here are a couple of improvements that could be made to the `evil-scroll-line-up` and `..-down` functions.

### Handle a 0 argument
**problem**:  A `0` argument stops C-e and C-y from scrolling.
**solution**: A `0` argument could reset the `evil-scroll-line-count` variable to `1`. 
Then it would be similar to how a `0` argument resets the `evil-scroll-up` / `-down` functions, to scroll by half the window height.

### Handle a negative argument
**problem**:  A negative argument scrolls in the reverse direction.
**solution**: Change a negative argument to positive.
**notes**: If there are situations where it would be useful to be able to scroll in the opposite direction, with a negative argument, then this could be left out.

### Only define duplicate lines once
**problem**:           6 lines in the "improved" `evil-scroll-line-up` and `..-down` functions below are identical.
**possible solution**: Maybe the scrolling commands could be defined in one of the `up` or `down` functions and the other one calls the first with a negative count argument.
Or there could be a "helper" function, that defines the duplicate lines once. It can then be called from both the `up` and `down` functions, with and without a negative count argument.
Or should a "helper" function just check the count argument and return it's new value?
**question**:          What should such a function be called? `evil-scroll-by-lines`?

### Add &optional before the count argument
**small issue**:  The count argument can be interpreted as being required. 
**solution**:     Add `&optional` before it.

### Scroll name confusion
**additional notes**: The scroll names are a little confusing, `evil-scroll-line-up` calls the Emacs `scroll-down` function, and `evil-scroll-line-down` calls the Emacs `scroll up` function.

They seem to be talking about scrolling the **window** `evil-scroll-line-up`:
>  "Scrolls the window COUNT lines upwards.

vs scrolling the **text** `scroll-down`:
>Scroll text of selected window down ARG lines.

Should a comment be added in the code about the difference? Or is there another way to reduce/prevent the risk of confusion.

### Docstring update
The docstrings was slightly rewritten and the argument outcomes added.

### Before
Currently the `evil-scroll-line-up` / `..-down` functions look like this:
```elisp
(evil-define-command evil-scroll-line-up (count)
  "Scrolls the window COUNT lines upwards.
If COUNT is not specified the function uses
`evil-scroll-line-count', which is the last used count."
  :repeat nil
  :keep-visual t
  (interactive "<c>")
  (progn
    (setq count (or count evil-scroll-line-count))
    (setq evil-scroll-line-count count)
    (let ((scroll-preserve-screen-position nil))
      (scroll-down count))))

(evil-define-command evil-scroll-line-down (count)
  "Scrolls the window COUNT lines downwards.
If COUNT is not specified the function uses
`evil-scroll-line-count', which is the last used count."
  :repeat nil
  :keep-visual t
  (interactive "<c>")
  (progn
    (setq count (or count evil-scroll-line-count))
    (setq evil-scroll-line-count count)
    (let ((scroll-preserve-screen-position nil))
      (scroll-up count))))
```

### After
With the possible improvements they would look like this (without a remove duplicates "helper" function):
```elisp
(evil-define-command evil-scroll-line-up (&optional count)
  "Scrolls the window COUNT lines upwards.
If COUNT isn't specified, then the last used COUNT
is read from the `evil-scroll-line-count' variable.
If COUNT is negative, then it's changed to positive.
If COUNT is 0, then it's reset to 1."
  :repeat nil
  :keep-visual t
  (interactive "<c>")
  (let ((count (cond ((not count)   evil-scroll-line-count)
                     ((> count 0)   count)
                     ((< count 0)   (* count -1))
                     ((zerop count) 1)))
        (scroll-preserve-screen-position nil))
    (setq evil-scroll-line-count count)
    (scroll-down count)))

(evil-define-command evil-scroll-line-down (&optional count)
  "Scrolls the window COUNT lines downwards.
If COUNT isn't specified, then the last used COUNT
is read from the `evil-scroll-line-count' variable.
If COUNT is negative, then it's changed to positive.
If COUNT is 0, then it's reset to 1."
  :repeat nil
  :keep-visual t
  (interactive "<c>")
  (let ((count (cond ((not count)   evil-scroll-line-count)
                     ((> count 0)   count)
                     ((< count 0)   (* count -1))
                     ((zerop count) 1)))
        (scroll-preserve-screen-position nil))
    (setq evil-scroll-line-count count)
    (scroll-up count)))
```

Any corrections or improvements are welcome.

**Comment by TheBB** [comment]
For the zero argument, sounds like a good idea. I don't see a reason to force arguments to be positive, it 
seems to me to violate PLOS. 

As for your other suggestions, I don't mind them, but let's not make a mountain of a mole hill. The code duplication in particular is not a problem here.

Lastly it looks like this should have been a pull request rather than an issue. Please do that.

**Comment by duianto** [comment]
I'm not familiar with PLOS, what does it stand for?

I'll remove the count begin forced to be positive. The duplicate lines and scroll name suggestions, were just observations so they are not important.

I noticed that `zerop` just is a macro of `(= count 0)`, so i'll change that as well.

I'll make a PR with the changes.

I thought that posting it as an issue first, would make it easier to find any obvious mistakes, and i would get input on the suggestions before i made a PR.

**Comment by TheBB** [comment]
That's fine. A PR is good for discussion too, though.

PLOS = principle of least surprise, at least I thought that worked in my head when I write it. I guess the “official” term is [principle of least astonishment](https://www.wikiwand.com/en/Principle_of_least_astonishment). Basically, it would be surprising to me if a function that accepted a count didn't work inversely when called with a negative argument, especially when the inverse operation is well defined.

**Comment by duianto** [comment]
Your right PLOS definitely applies, if one enters a negative count then one would expect it to scroll in the opposite direction.

The next time when i have some working code for a suggestion, then I'll post it as a PR. And add any questions in the PR's description.

**Comment by wasamasa** [comment]
Closed in favor of #791.
